xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -315,507 +322,948 @@ func setupWithWatch(t *testing.T) (*fakeclient.Client, *cdsBalancer, *testEDSBal
-			wantCCS: edsCCS(serviceName, nil, false, wrrLocalityLBConfigJSON, json.RawMessage(`{
-				"interval": "10s",
-				"baseEjectionTime": "30s",
-				"maxEjectionTime": "300s",
-				"maxEjectionPercent": 10,
-				"successRateEjection": {
-					"stdevFactor": 1900,
-					"enforcementPercentage": 100,
-					"minimumHosts": 5,
-					"requestVolume": 100
-				},
-				"failurePercentageEjection": {
-					"threshold": 85,
-					"enforcementPercentage": 5,
-					"minimumHosts": 5,
-					"requestVolume": 50
-				}
-			}`)),
 		},
 	}
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-			defer ctxCancel()
-			if err := invokeWatchCbAndWait(ctx, xdsC, cdsWatchInfo{test.cdsUpdate, test.updateErr}, test.wantCCS, edsB); err != nil {
+			lbCfgCh, _, _, _ := registerWrappedClusterResolverPolicy(t)
+			mgmtServer, nodeID, _, _, _, _, _ := setupWithManagementServer(t)
+
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			if err := mgmtServer.Update(ctx, e2e.UpdateOptions{
+				NodeID:         nodeID,
+				Clusters:       []*v3clusterpb.Cluster{test.clusterResource},
+				SkipValidation: true,
+			}); err != nil {
+				t.Fatal(err)
+			}
+
+			if err := compareLoadBalancingConfig(ctx, lbCfgCh, test.wantChildCfg); err != nil {
 				t.Fatal(err)
 			}
 		})
 	}
 }
 
-// TestHandleClusterUpdateError covers the cases that an error is returned from
-// the watcher.
-func (s) TestHandleClusterUpdateError(t *testing.T) {
-	// This creates a CDS balancer, pushes a ClientConnState update with a fake
-	// xdsClient, and makes sure that the CDS balancer registers a watch on the
-	// provided xdsClient.
-	xdsC, cdsB, edsB, tcc, cancel := setupWithWatch(t)
-	defer func() {
-		cancel()
-		cdsB.Close()
-	}()
-
-	// A watch was registered above, but the watch callback has not been invoked
-	// yet. This means that the watch handler on the CDS balancer has not been
-	// invoked yet, and therefore no EDS balancer has been built so far. A
-	// resolver error at this point should result in the CDS balancer returning
-	// an error picker.
-	watcherErr := errors.New("cdsBalancer watcher error")
-	xdsC.InvokeWatchClusterCallback(xdsresource.ClusterUpdate{}, watcherErr)
-
-	// Since the error being pushed here is not a resource-not-found-error, the
-	// registered watch should not be cancelled.
-	sCtx, sCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
-	defer sCancel()
-	if _, err := xdsC.WaitForCancelClusterWatch(sCtx); err != context.DeadlineExceeded {
-		t.Fatal("cluster watch cancelled for a non-resource-not-found-error")
+// Tests a single success scenario where the cds LB policy receives a cluster
+// resource from the management server with LRS enabled. Verifies that the load
+// balancing configuration pushed to the child is as expected.
+func (s) TestClusterUpdate_SuccessWithLRS(t *testing.T) {
```

Comment:
I'm sure there's a reason, but why is this not merged in with the t test directly above?

Reply:
I tried to merge them but it made the other test way more complicated than it had to. The reason is that the lrs config in the resource from the management server simply has a config source which says "self". But this has to be translated into the server config from the bootstrap. This means that the server URI (in the wantConfig) has to depend on the actual address of the management server. This means that I can't statically define wantConfig in the test table in the previous test.

Reply:
Hmm ok. I think you could argue you could have a conditional and write an extra field to the static configuration. I.e. if test name {test.wantChildConfig.DiscoveryMechanism.LoardReportingServer = want}. But this isn't too many lines of code so seems fine :).